File: 1036-RottingOranges/1036-RottingOranges.java
// Last updated: 8/16/2026, 9:49:03 PM
class Solution {
    public int orangesRotting(int[][] grid) {
        // BFS
        if (grid.length == 0) return 0;

        int m = grid.length, n = grid[0].length;
        int[][] dirs = {{0, 1}, {0, -1}, {1, 0}, {-1, 0}};
        int res = 0;

        int cntFresh = 0;
        Queue<int[]> queue = new LinkedList<>();
        for (int i = 0; i < m; i ++) {
            for (int j = 0; j < n; j ++) {
                if (grid[i][j] == 1) cntFresh ++;
                // add all rotten to queue, rotten also means visited mark
                if (grid[i][j] == 2) queue.offer(new int[]{i, j});
            }
        }

        if (cntFresh == 0) return 0;

        // BFS starts from initially rotten oranges
        while (!queue.isEmpty()) {
            int qSize = queue.size();
            // System.out.println(queue.size());
            for (int p = 0; p < qSize; p ++) {
                int[] curr = queue.poll();
                // System.out.println(Arrays.toString(curr));
                int status = grid[curr[0]][curr[1]];
                
                for (int[] dir: dirs) {
                    int dx = curr[0] + dir[0];
                    int dy = curr[1] + dir[1];
                    if (dx < 0 || dx >= m || dy < 0 || dy >= n) continue;
                    if (grid[dx][dy] != 1) continue;
                    queue.offer(new int[]{dx, dy});
                    grid[dx][dy] = 2; // 2 means visited/rotten
                    cntFresh --; // update current fresh count
                }             
            }
            res++;
        }
        // System.out.println(Arrays.deepToString(grid));
        return (cntFresh == 0) ? res - 1 : -1;
    }
}

File: 1036-RottingOranges/1036-RottingOranges.py
# Last updated: 8/16/2026, 9:48:59 PM
from collections import deque

class Solution:
    def orangesRotting(self, grid: List[List[int]]) -> int:
        # multi-source BFS: level traversal

        m, n = len(grid), len(grid[0])
        q = deque()
        fresh = 0

        # s-1: enqueue all rotten oranges
        for i in range(m):
            for j in range(n):
                if grid[i][j] == 2:
                    q.append((i, j))
                if grid[i][j] == 1:
                    fresh += 1
        time = 0

        # fresh > 0 is required, otherwise, we will traverse one more round, and needs return time-1
        # also when no fresh oranges left, thats the right moment we need to calcualte the time
        while q and fresh > 0:
            size = len(q)
            for _ in range(size):
                i, j = q.popleft()
                for dx, dy in [[-1, 0], [1, 0], [0, -1], [0, 1]]:
                    x = i + dx
                    y = j + dy
                    if 0 <= x < m and 0 <= y < n and grid[x][y] == 1:
                        grid[x][y] = 2 # mark as visited during enqueue
                        fresh -= 1
                        q.append((x, y))
            # print(q)
            time += 1
        
        return time if fresh == 0 else -1